log-replication/src/main/java/org/corfudb/logreplication/fsm/InSnapshotSyncState.java

```diff
@@ -0,0 +1,200 @@
+
+    /*
+     Read and send a snapshot of the data-store.
+     */
+    @Getter
+    @VisibleForTesting
+    private SnapshotSender snapshotSender;
+
+    /*
+     A future on the send, in case we need to cancel the ongoing snapshot sync.
+     */
+    private Future<?> transmitFuture;
+
+    /**
+     * Constructor
+     *
+     * @param logReplicationFSM log replication state machine
+     * @param snapshotSender snapshot sync send (read and send)
+     */
+    public InSnapshotSyncState(LogReplicationFSM logReplicationFSM, SnapshotSender snapshotSender) {
+        this.fsm = logReplicationFSM;
+        this.snapshotSender = snapshotSender;
+    }
+
+    @Override
+    public LogReplicationState processEvent(LogReplicationEvent event) throws IllegalTransitionException {
+        switch (event.getType()) {
+            case SNAPSHOT_SYNC_REQUEST:
+                /*
+                 Cancel ongoing snapshot sync, if it is still in progress.
+                 */
+                cancelSnapshotSync("another snapshot sync request.");
+
+                /*
+                 Set the id of the new snapshot sync request causing the transition.
+
+                 This will be taken onEntry of this state to initiate a snapshot send
+                 for this given request.
+                 */
+                setTransitionEventId(event.getEventID());
+                snapshotSender.reset();
+                return this;
+            case SNAPSHOT_SYNC_CONTINUE:
+                /*
+                 Snapshot sync is broken into multiple tasks, where each task sends a batch of messages
+                 corresponding to this snapshot sync. This is done to accommodate the case
+                 of multi-site replication sharing a common thread pool, continuation allows to send another
+                 batch of updates for the current snapshot sync.
+                 */
+                if (event.getMetadata().getRequestId() == transitionEventId) {
+                    log.debug("Continuation of snapshot sync for {}", event.getEventID());
+                    return this;
+                } else {
+                    log.warn("Unexpected snapshot sync continue event {} when in snapshot sync state {}.",
+                            event.getEventID(), transitionEventId);
+                }
+            case SYNC_CANCEL:
+                // If cancel was intended for current snapshot sync task, cancel and transition to new state
+                if (transitionEventId == event.getMetadata().getRequestId()) {
+                    cancelSnapshotSync("cancellation request.");
+                    // Re-trigger SnapshotSync due to error, generate a new event Id for the new snapshot sync
+                    LogReplicationState inSnapshotSyncState = fsm.getStates().get(LogReplicationStateType.IN_SNAPSHOT_SYNC);
+                    inSnapshotSyncState.setTransitionEventId(UUID.randomUUID());
+                    snapshotSender.reset();
```

Comment:
should it not go back into sync through the event queue. to me this seems like a 
in_sync -> some state like init -> in_sync.  what do you think.